src/app/DeferredAttributePersistenceProvider.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/DeferredAttributePersistenceProvider.h>
+
+#include <platform/CHIPDeviceLayer.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR DeferredAttribute::PrepareWrite(AttributePersistenceProvider & persister, const EmberAfAttributeMetadata * metadata,
+                                           const ByteSpan & value)
+{
+    mPersister = &persister;
+    mMetadata  = metadata;
+
+    if (mValue.AllocatedSize() != value.size())
+    {
+        mValue.Alloc(value.size());
+        ReturnErrorCodeIf(!mValue, CHIP_ERROR_NO_MEMORY);
+    }
+
+    memcpy(mValue.Get(), value.data(), value.size());
+    return CHIP_NO_ERROR;
+}
+
+void DeferredAttribute::Flush()
+{
+    VerifyOrReturn(mValue);
+    mPersister->WriteValue(mPath, mMetadata, ByteSpan(mValue.Get(), mValue.AllocatedSize()));
+    mValue.Release();
+}
+
+CHIP_ERROR DeferredAttributePersistenceProvider::WriteValue(const ConcreteAttributePath & path,
+                                                            const EmberAfAttributeMetadata * metadata, const ByteSpan & value)
+{
+    for (DeferredAttribute & da : mDeferredAttributes)
+    {
+        if (da.Matches(path))
+        {
+            ReturnErrorOnFailure(da.PrepareWrite(mPersister, metadata, value));
+            DeviceLayer::SystemLayer().StartTimer(System::Clock::Milliseconds32(mWriteDelay), Flush, &da);
```

Comment:
So this will have a separate timer running per attribute, right?  Would it make any sense to have a single timer running if any of the attributes are dirty, then when it fires flush them all?  That would allow us to remove the persister member from `DeferredAttribute`, and might put less pressure on the timer pool...

The fact that the delay is global for the persister but the timers are per-deferred-attribute needs to at least be documented, beause it's very non-obvious.

Reply:
My goal was to make sure that if someone calls `Move-To-Level` or any other long-running command that changes an attribute, that attribute is persisted only if the command. So each new write restarts the timer. I wasn't sure if it's OK to restart the flush timer for attribute A when changing attribute B. If it is, I can change it. I can also add timeout member to `DeferredAttribute` and have the timer check which attributes can be written at a given time.

Reply:
Ah, I see.  Yes, if we're trying to not write until the attribute goes entirely quiescent (compared to our interval), then this setup makes sense.  I thought we were just trying to throttle.

We should probably clearly document in the API that an attribute will not be written out at all if it never stops changing for the timeout duration...

Or if we're OK with the throttling approach we can just avoid resetting the timer if it's already set and then flush all attributes when the timer fires.  I don't have strong views on which is better; I do think we should have the API clearly document what the behavior is.

Reply:
OK, I'll think about it over the weekend and decide on the either approach :) Thanks for the feedback.

Reply:
I documented the current behavior, but I also changed the code to use a single timer no matter how many attributes have writes deferred.